core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1297,7 +1299,10 @@ class ReplicaManager(val config: KafkaConfig,
             }
           }
         }
+        // handle the new partition
         replicaAlterLogDirsManager.addFetcherForPartitions(futureReplicasAndInitialOffset)
+        // handle the partitions having new epoch
+        replicaAlterLogDirsManager.updateEpoch(updatedPartitions)
```

Comment:
For the normal fetcher, we force the replica into a truncating phase after each epoch change. I'm trying to convince myself whether we need to do this for the log dir fetcher or not. Although we have similar log reconciliation logic implemented for the log dir fetcher, it seems we only exercise it when the log dir fetcher is first initialized for a partition. Outside of that, we have an asynchronous truncation path which gets executed after the current log is truncated upon becoming follower. If we think this is sufficient to guarantee consistency, then I'm wondering whether the epoch validation when the log dir fetcher is fetching against the active log is actually buying us anything. An alternative would be that we leave the epoch uninitialized which would cause us to skip epoch validation.

@junrao Any thoughts?

Reply:
@hachikuji : Another thing is that epoch could potentially help in the case of unclean leader election.

Reply:
I gave this a bit more thought. I think it _would_ be safer to have the log dir fetcher mimic the regular fetcher when it comes to the truncation protocol. That said, I don't think we need to address this here. The fix in this patch addresses the main problem at the moment.